File: 0547-number-of-provinces/0547-number-of-provinces.java
class Solution {

    public void BFS(int[][] adj, int u, boolean[] visited){
       Queue<Integer> que = new LinkedList<>();
       que.offer(u);
       while(!que.isEmpty()){
            int t = que.poll();
            visited[t] = true;
            for(int v = 0; v < adj.length; v++){
                if(!visited[v] && adj[t][v] == 1){
                    que.offer(v);
                }
            }
       }

    }
    public int findCircleNum(int[][] isConnected) {
       int n = isConnected.length;
        boolean[] visited = new boolean[n];
        
        int provinces = 0;
        // call DFS and count provinces
        for(int u = 0; u < n; u++){
            if(!visited[u]){
                provinces++;
                BFS(isConnected, u, visited);
            }
        }
        return provinces;
    }
}

File: 0547-number-of-provinces/0547-number-of-provinces.cpp
class Solution {
public:
    void DFS(vector<vector<int>>& adj, int u, vector<bool> &visited){
        
        visited[u] = true;
        for(int v = 0 ; v < adj[u].size(); v++){
            if(!visited[v] && adj[u][v] == 1){
                DFS(adj, v, visited);
            }
        } 
        
    }
    int findCircleNum(vector<vector<int>>& adj) {
        // make adjancecy list
        int n = adj.size();

        vector<bool> visited(n, false);
        int countOfProvinces = 0;
        for(int u = 0 ; u<n; u++){
            if(!visited[u]){
                DFS(adj, u, visited);
                countOfProvinces++;
            }
        }

        return countOfProvinces;
    }
};